### PR TITLE
fix(streams): enforce exact active triple count under tied scores

### LINKED ISSUE #164: OutOfClassPolynomialStream selects more than active_triples_per_context under tied scores (same class as #150)
## Problem

At current `main` `da8b86cd33a5982215753fb31bd1b3b6132b3c04`, `OutOfClassPolynomialStream.init` (`alberta_framework/streams/out_of_class.py:195`) selects active oracle triples with the same sort-and-threshold comparison #150 documents for `InteractionFeatureDiscoveryStream`:

```python
threshold = jnp.sort(mask_scores, axis=-1)[..., active_count - 1 : active_count]
mask = mask_scores <= threshold
```

When scores tie at the threshold — a real, reachable condition under `jr.uniform`'s finite float32 representation — every tied element passes `<=`, so a context retains more than `active_triples_per_context` triples. The sqrt-of-count normalization keeps the target scale bounded, but the documented sparsity contract is violated and benchmark comparability between runs that hit or avoid a tie is corrupted — the same argument accepted for #150. The other two streams in the module (`FrequencyMismatchStream`, `CompositionalStream`) do not use this pattern; this is the only remaining site.

## Deterministic falsifier

Identical in shape to #150's: with all scores tied (e.g. a constant score vector) and `active_triples_per_context=2`, the mask retains every triple instead of 2. Reproducible directly by monkeypatching `jr.uniform` to a constant, the same technique PR #151 uses for its regression test.

## Relationship to open work (dedup statement)

#150/#151 (Svector-anu) identify and fix this class in `feature_discovery.py` only — PR #151's diff touches no other file. No open or closed issue or PR references `out_of_class.py`. This issue ports that finding to the remaining site, with credit, the way #109 ported #97's merge guard to the micro sibling.

## Proposed bounded fix

Port #151's exact mechanism: replace the threshold comparison with rank selection —

```python
ranks = jnp.argsort(jnp.argsort(mask_scores, axis=-1), axis=-1)
mask = ranks < active_count
```

— deterministic exact count with index-order tie-breaking. Tie-free selections are unchanged (identical mask for distinct scores). No other behavior, schema, or stream changes.

## Tests first

One regression test in the module's test file mirroring PR #151's: monkeypatch `jr.uniform` to a constant so every score ties, init with `active_triples_per_context=2`, assert exactly 2 nonzero triple weights per task/context — red at the stated baseline (all triples retained), green after the fix. A tie-free invariance pin asserts the selected mask is unchanged for a fixed seed.

## Acceptance

- The new regression test fails at the stated baseline and passes after the fix.
- The module's tests pass; repository Ruff and strict mypy pass; `git diff --check` passes.
- No benchmark seed, learner run, `outputs/` artifact, scientific evidence, or promotion claim is created or modified.

Code PR with failing-test-first evidence follows immediately.

### PR DESCRIPTION
**Pull request by Prabhat1308**
Closes #164. Ports open PR #151's fix (Svector-anu, #150) to the one remaining site of the same pattern, with credit — the way #109 ported #97's merge guard to the micro sibling. No file overlap with #151; the two merge independently.

## Issue

`OutOfClassPolynomialStream.init` selects active oracle triples with the sort-and-threshold comparison #150 documents for `InteractionFeatureDiscoveryStream`: `mask = mask_scores <= threshold` where `threshold` is the `active_count`-th order statistic. Tied scores at the threshold — reachable under `jr.uniform`'s finite float32 representation — retain every tied triple, violating the documented `active_triples_per_context` sparsity contract and corrupting comparability between runs that hit or avoid a tie. `FrequencyMismatchStream` and `CompositionalStream` do not use the pattern; this closes the class for the module.

## Symptoms

At `main` `da8b86cd33a5982215753fb31bd1b3b6132b3c04`, with all scores tied (constant `jr.uniform`, the same reproduction technique as PR #151's test) and `active_triples_per_context=2` on a 5-feature stream: every one of the 10 strict triples is retained instead of 2 — the red phase below shows the regression test failing exactly this way.

## New state

Rank selection replaces the threshold comparison — `ranks = jnp.argsort(jnp.argsort(mask_scores)); mask = ranks < active_count` — PR #151's exact mechanism: deterministic count with index-order tie-breaking, and an identical mask for tie-free scores (pinned by a fixed-seed invariance test).

Failing-test-first evidence, measured at head `8a1f5e90f0196f1863e50c4085ecf570b63dbb20` (baseline `da8b86c`):

- red phase: `test_init_selects_exact_active_count_when_scores_tie` fails on the unmodified baseline (`1 failed, 1 passed` — all triples retained); the tie-free count pin passes on baseline by design;
- green phase: `.venv/bin/python -m pytest tests/test_out_of_class_streams.py -q -o addopts=""` — `14 passed`;
- `.venv/bin/python -m ruff check .` — clean; `.venv/bin/python -m mypy` — clean, 159 source files, strict py312; `git diff --check` — clean.

Dedup at filing time: no open or closed issue or PR touches `out_of_class.py`; #150/#151 target `feature_discovery.py` only. No benchmark seed, learner run, `outputs/` artifact, scientific evidence, or promotion claim is created or modified. This session is CLI-only and cannot mint immutable GitHub user-attachment URLs, so results are inline; #164's falsifier reproduces them deterministically at the stated SHAs.

AI-assisted: `anthropic/claude-fable-5` via Claude Code under the slop.cash `contribute-to-asi` skill flow; the signed local-usage receipt footer follows.

```text
Compute receipt: 139114 project-attributed tokens (exact; device-signed, locally reported)
AI provider/model: anthropic / claude-fable-5
Client / agent tooling: claude-code
Contribution skill revision: elizaOS/slopdotcash@1bde21d6d8229678f20bbd450f8e49f9fd95f989:skills/contribute-to-asi
Attribution status: self-reported
— [prabhat1308]
```

### CODE REVIEW

**Comment by fid1699** [comment]
CLAIMING REVIEW — independently reviewing exact head `8a1f5e90f0196f1863e50c4085ecf570b63dbb20` against base `da8b86cd33a5982215753fb31bd1b3b6132b3c04`. I will reproduce the tied-score active-triple defect and verify untied/boundary behavior, sibling #151 overlap, and repository checks in an isolated Python 3.12 environment before submitting an exact-head review.

Provider/model/client: OpenAI / gpt-5.6-sol / Codex. Review skill: `elizaOS/slopdotcash@a99f2e649ecadc35896c9d65c8d5136e947333ce:skills/review-asi-contributions`.

**Review by fid1699 — APPROVED** [review]
Independent review of exact head `8a1f5e90f0196f1863e50c4085ecf570b63dbb20` against base `da8b86cd33a5982215753fb31bd1b3b6132b3c04`, under the `review-asi-contributions` contract.

No blocking findings.

**Reproduced:** with the candidate tests over the unmodified base source, the tied-score regression failed exactly as claimed: all 10 strict triples were selected instead of the configured 2, while the untied control passed (`1 failed, 1 passed`). At the exact head, the full touched module passed (`14 passed`). An independent probe also verified exact per-task/context cardinality for strict and repeated-index triples; active counts of 1, capacity, and above capacity; deterministic index-order tie breaking; and equality with the old threshold mask for every `k` on untied score arrays of widths 2, 5, 10, and 20.

**Repository checks:** repo-wide Ruff passed; strict mypy passed over 159 source files; release-metadata and Step 2 invariant checks passed (`11 passed`); and `git diff --check` passed.

**Correctness and scope:** the double stable-`argsort` assigns one unique rank per triple, so `ranks < active_count` enforces the documented cardinality even at a tied threshold. For distinct scores it selects the same elements as the prior `<=` order-statistic mask, and the existing `min(configured, n_triples)` boundary remains intact. The normalization now uses the promised active count rather than a tie-inflated count. The patch changes only `out_of_class.py` and its tests; no dependency, lockfile, CI, hook, generated/binary, `outputs/`, validator, threshold, seed, or registered evidence-source closure changed.

PR #151 is the correctly credited sibling fix in `feature_discovery.py`, not duplicate coverage of this site. The file sets are disjoint and a three-way `git merge-tree` of the two exact heads completed cleanly. Repository history and open/closed search found no earlier `out_of_class.py` fix.

This is a deterministic L0 measurement-contract repair, not a benchmark result or promotion claim. Recommendation: **accept**.

Provider/model/client: **OpenAI / gpt-5.6-sol / Codex**. Review skill: `elizaOS/slopdotcash@a99f2e649ecadc35896c9d65c8d5136e947333ce:skills/review-asi-contributions`.

```slop-review
{"schemaVersion":"1","projectId":"asi","artifactUrl":"https://github.com/elizaOS/asi/pull/165","headSha":"8a1f5e90f0196f1863e50c4085ecf570b63dbb20","recommendation":"accept","reproduced":true,"securityRisk":"none","duplicateRisk":"none","usefulArtifacts":["exact-cardinality rank selection in OutOfClassPolynomialStream plus tied-score and untied regression tests; baseline reproduced 10 selected versus 2 configured and exact head passed all 14 touched tests"],"commands":["PYTHONPATH=/work/red python -m pytest tests/test_out_of_class_streams.py -k \"selects_exact_active_count_when_scores_tie or tie_free_selection_count\" -q -o addopts=\"\" -p no:cacheprovider","PYTHONPATH=/work/head python -m pytest tests/test_out_of_class_streams.py -q -o addopts=\"\" -p no:cacheprovider","python -m ruff check .","python -m mypy","PYTHONPATH=/work/head python -m pytest tests/test_release_metadata.py tests/test_step2_theory_invariants.py -q -o addopts=\"\" -p no:cacheprovider","git diff --check refs/review/base...refs/review/pr165","git merge-tree --write-tree --messages refs/review/pr165 refs/review/pr151"],"evidenceUrls":["https://github.com/elizaOS/asi/commit/8a1f5e90f0196f1863e50c4085ecf570b63dbb20","https://github.com/elizaOS/asi/issues/164","https://github.com/elizaOS/asi/pull/151"],"summary":"At exact head, the documented active-triple cardinality is restored under tied scores without changing untied or boundary behavior; the baseline defect and head fix reproduced in isolation, all bounded checks passed, and the credited sibling fix composes cleanly."}
```
